### DISCUSSION TITLE
Will Blazor United let me do this?

### DISCUSSION THREAD

**Discussion by ThatRendle** [dotnet/aspnetcore#49027] [Q&A] [answered]
I just want to find out whether I've understood Blazor United right...

Will I be able to build a web app that:

- Can be fully server-side rendered for search engines, browsers without JS, etc.
- Will serve SSR pages while the Blazor WebAssembly files are being downloaded in the background
- And then switch seamlessly to SPA mode allowing me to override links and provide interactivity using standard Blazor code
  - (Like SvelteKit or Astro or similar)
- Makes me peanut butter and banana sandwiches whenever I'm a bit peckish

**Comment by SteveSandersonMS** [comment, marked as answer]
I think the answer is yes, though I'd be interested to clarify your exact requirements more.

The things you're describing were largely already achievable in .NET 7 and earlier. That is, a Blazor WebAssembly hosted app could be prerendered on the server, and then:

 * It will serve HTML and support link-based navigation without needing any `.js` file to be loaded (or any circuit/wasm runtime)
 * If you did reference `blazor.webassembly.js`, it would download the runtime in the background and then start up interactive WebAssembly-based behaviors once the runtime was downloaded, seamlessly turning on SPA-like behaviors such as intercepting navigation to perform it on the client

What's better for your specific situation with .NET 8 is:

 * The server-side rendering parts are much cleaner to set up and integrate more nicely with ASP.NET Core endpoint routing
 * Server-side request handling can deal with POST requests (model binding) as well as just GET requests, so you can achieve some interactive behaviors like forms without depending on `.js`/circuits/wasm. So it's not only limited to people clicking links before the wasm runtime starts.
 * Until the WebAssembly runtime starts, the site will operate in a progressively-enhanced but non-wasm/circuit based mode. That is, if you reference `blazor.web.js` but don't mark any components as Server/WebAssembly (or you did pick WebAssembly but it's still in process of downloading the runtime files), you get a stateless HTML/HTTP site, but there's some client-side code that intercepts navigation and form posts and resolves it via `fetch` requests to give a smoother experience (like Turbolinks) and not interrupt the download process for the wasm runtime while the user is navigating around.
 * There may be parts (e.g., pages) of your site that don't need WebAssembly at all. For those, you can now simply leave them as static pages, and the user won't even have to start downloading WebAssembly files until they navigate to somewhere that needs it, reducing bandwidth costs for both server and client.

Obviously there are many other enhancements too in .NET 8, but these are the ones I think directly relate to the scenario you described.

Does that make sense? I'm hoping it will be clear that most of what you're asking was technically already possible in .NET 7 (i.e., a prerendered wasm site) and so the most relevant thing to distinguish is what you're looking for that wasn't already supported in .NET 7.

> Makes me peanut butter and banana sandwiches whenever I'm a bit peckish

Only one ingredient is supported in .NET 8, but if there's sufficient demand we will consider supporting arbitrarily many (at least up to `int.MaxValue`) in a future update.

**Reply by ThatRendle** [reply]
Thanks Steve! I knew it was basically possible in .NET 7 but the unified/simplified model in 8 makes it much more attractive IMO.

I've got a version of the site running as purely ASP.NET Core MVC right now, with a tiny amount of JS-provided progressive enhancement. I'm planning on Twitch/YouTube streaming the migration to the new Blazor whenever the preview bits with WASM support land.
